FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/DicomToDataLakeProcessingJob.cs

```diff
@@ -165,8 +170,43 @@ public async Task<string> ExecuteAsync(IProgress<string> progress, CancellationT
 
             if (limit > 0)
             {
-                var changeFeedOptions = new ChangeFeedOptions(_inputData.StartOffset, limit, false);
-                var metadataList = await _dataClient.GetMetadataAsync(changeFeedOptions, cancellationToken);
+                // Get change feeds
+                var queryParameters = new List<KeyValuePair<string, string>>
+                {
+                    new KeyValuePair<string, string>(DicomApiConstants.OffsetKey, $"{_inputData.StartOffset}"),
+                    new KeyValuePair<string, string>(DicomApiConstants.LimitKey, $"{limit}"),
+                    new KeyValuePair<string, string>(DicomApiConstants.IncludeMetadataKey, $"{false}"),
+                };
+
+                var changeFeedOptions = new ChangeFeedOffsetOptions(dicomVersion, queryParameters);
+                string changeFeedsContent = await _dataClient.SearchAsync(changeFeedOptions, cancellationToken);
+
+                List<ChangeFeedEntry> changeFeedEntries;
+                try
+                {
+                    changeFeedEntries = JsonConvert.DeserializeObject<List<ChangeFeedEntry>>(changeFeedsContent)
+                        .Where(x => x.State == ChangeFeedState.Current && x.Action == ChangeFeedAction.Create)
+                        .ToList();
+                }
+                catch (Exception ex)
+                {
+                    _diagnosticLogger.LogError($"Parse changefeeds failed. Reason: {ex.Message}");
+                    _logger.LogError($"Parse changefeeds failed. Reason: {ex.Message}");
+                    throw new FhirSearchException("Parse changefeeds failed.", ex);
+                }
+
+                // Get metadata list
+                var metadataList = new List<string>();
+                foreach (var changeFeedEntry in changeFeedEntries)
+                {
+                    var metadataOptions = new SearchMetadataOptions(
+                        dicomVersion,
+                        changeFeedEntry.StudyInstanceUid,
+                        changeFeedEntry.SeriesInstanceUid,
+                        changeFeedEntry.SopInstanceUid);
+
+                    metadataList.Add(await _dataClient.SearchAsync(metadataOptions, cancellationToken));
+                }
```

Comment:
We can consider search them concurrently for better performance.

Reply:
Updated, thanks.